modules/3d/src/rendering.cpp

```diff
@@ -0,0 +1,333 @@
+            if ((f[0] >= 0) && (f[1] >= 0) && (f[2] >= 0))
+            {
+                bool update = false;
+                float zInter = 0.f;
+                if (!depthBuf.empty())
+                {
+                    zInter = 1.0f / (f[0] * invz[0] + f[1] * invz[1] + f[2] * invz[2]);
+                    float zCurrent = depthBuf.at<float>(y, x);
+                    if (zInter < zCurrent)
+                    {
+                        update = true;
+                        depthBuf.at<float>(y, x) = zInter;
+                    }
+                }
+                else // Shading::White
+                {
+                    update = true;
+                }
+
+                if (!colorBuf.empty() && update)
+                {
+                    Vec3f color {0, 0, 0};
+                    if (settings.shadingType == ShadingType::White)
+                    {
+                        color = { 1.f, 1.f, 1.f };
+                    }
+                    if (settings.shadingType == ShadingType::Flat)
+                    {
+                        color = colors[0];
+                    }
+                    else // ShadingType::Shaded
+                    {
+                        Vec3f interp;
+                        for (int i = 0; i < 3; i++)
+                        {
+                            interp[i] = f[i] * zInter * invz[i];
+                        }
+
+                        for (int i = 0; i < 3; i++)
+                        {
+                            for (int j = 0; j < 3; j++)
+                            {
+                                color[i] += interp[j] * colors[j][i];
+                            }
+                        }
+                    }
+                    colorBuf.at<Vec3f>(y, x) = color;
+                }
+            }
+        }
+    }
+}
+
+
+CV_EXPORTS  void triangleRasterize(InputArray _vertices, InputArray _indices, InputArray _colors,
+                                   InputArray cameraPose, float fovyRadians, float zNear, float zFar,
+                                   int width, int height, RasterizeSettings settings,
+                                   OutputArray _depthBuffer, OutputArray _colorBuffer)
+{
+    bool needDepth = _depthBuffer.needed();
+    bool needColor = _colorBuffer.needed();
+
+    if (!needDepth && !needColor)
+    {
+        CV_Error(Error::StsBadArg, "No depth nor color output image provided");
+    }
+
+    CV_Assert(cameraPose.type() == CV_32F);
+    CV_Assert((cameraPose.size() == Size {4, 3}) || (cameraPose.size() == Size {4, 4}));
+
+    CV_Assert((fovyRadians > 0) && (fovyRadians < CV_PI));
+    CV_Assert(zNear > 0);
+    CV_Assert(zFar > zNear);
+    CV_Assert((width > 0) && (height > 0));
+
+    Mat cpMat = cameraPose.getMat();
+    Matx44f camPoseMat = Matx44f::eye();
+    for (int i = 0; i < 3; i++)
+    {
+        for (int j = 0; j < 4; j++)
+        {
+            camPoseMat(i, j) = cpMat.at<float>(i, j);
+        }
+    }
```

Comment:
copyTo/ConvertTo/initialize from .data ptr?

Reply:
Input matrix can be 4x3 or 4x4; this way of initialization seems to be the most simple and universal